crates/gossip/src/protocol.rs

```diff
@@ -1,26 +1,143 @@
+            let inner = K2GossipAcceptMessage::decode(outer.data)
+                .map_err(K2Error::other)?;
+            Ok(GossipMessage::Accept(inner))
+        }
+        k2_gossip_message::GossipMessageType::NoDiff => {
+            let inner = K2GossipNoDiffMessage::decode(outer.data)
+                .map_err(K2Error::other)?;
+            Ok(GossipMessage::NoDiff(inner))
+        }
+        k2_gossip_message::GossipMessageType::Agents => {
+            let inner = K2GossipAgentsMessage::decode(outer.data)
+                .map_err(K2Error::other)?;
+            Ok(GossipMessage::Agents(inner))
+        }
+        _ => Err(K2Error::other("Unknown gossip message type".to_string())),
+    }
 }
 
 /// Serialize a gossip message
 pub fn serialize_gossip_message(
-    value: K2GossipMessage,
+    value: GossipMessage,
 ) -> K2Result<bytes::Bytes> {
     let mut out = bytes::BytesMut::new();
 
-    value.encode(&mut out).map_err(|e| {
+    let (msg_type, inner) = serialize_inner_gossip_message(value)?;
+
+    K2GossipMessage {
+        msg_type: msg_type as i32,
+        data: inner,
+    }
+    .encode(&mut out)
+    .map_err(|e| {
         K2Error::other(format!("Failed to serialize gossip message: {:?}", e))
     })?;
 
     Ok(out.freeze())
 }
+
+fn serialize_inner_gossip_message(
+    value: GossipMessage,
+) -> K2Result<(k2_gossip_message::GossipMessageType, bytes::Bytes)> {
+    let mut out = bytes::BytesMut::new();
+
+    match value {
+        GossipMessage::Initiate(inner) => {
+            inner.encode(&mut out).map_err(|e| {
+                K2Error::other(format!(
+                    "Failed to serialize gossip message: {:?}",
+                    e
+                ))
+            })?;
+
+            Ok((k2_gossip_message::GossipMessageType::Initiate, out.freeze()))
+        }
+        GossipMessage::Accept(inner) => {
+            inner.encode(&mut out).map_err(|e| {
+                K2Error::other(format!(
+                    "Failed to serialize gossip message: {:?}",
+                    e
+                ))
+            })?;
+
+            Ok((k2_gossip_message::GossipMessageType::Accept, out.freeze()))
+        }
+        GossipMessage::NoDiff(inner) => {
+            inner.encode(&mut out).map_err(|e| {
+                K2Error::other(format!(
+                    "Failed to serialize gossip message: {:?}",
+                    e
+                ))
+            })?;
+
+            Ok((k2_gossip_message::GossipMessageType::NoDiff, out.freeze()))
+        }
+        GossipMessage::Agents(inner) => {
+            inner.encode(&mut out).map_err(|e| {
+                K2Error::other(format!(
+                    "Failed to serialize gossip message: {:?}",
+                    e
+                ))
+            })?;
+
+            Ok((k2_gossip_message::GossipMessageType::Agents, out.freeze()))
+        }
+    }
+}
+
+/// Encode agent ids as bytes
+pub(crate) fn encode_agent_ids(
```

Comment:
We could add this or `From<IntoIterator>` for ids to the `api/id` module. They're used in multiple places.

Reply:
Where else is sending agent ids over the wire?

Reply:
Not agent ids, but just ids, e.g. op ids. There's a `From<Bytes> for Id` and a `From<Id> for AgentId`.

Reply:
Oh I get you! Yeah that's a nice idea

Reply:
Done here -> https://github.com/holochain/kitsune2/pull/72/commits/e61dfbc3645efb8ebe2fcde43bf289bc8f56af2e